src/docs/arch-concurrency.dox

```diff
@@ -0,0 +1,107 @@
+
+@section locking Locking
+
+There are three types of locks in WiredTiger:
+ - Condition Variables
+ - Spin Locks
+ - Read Write (RW) Locks
+Each mechanism guarantees that once a thread relinquishes its lock and another thread obtains the
+same lock, all modifications made by the first thread will be visible to the thread now holding
+the lock.
+
+There are two implementations of condition variables in WiredTiger respectively for POSIX and MSVC
+under the same interface. The specific implementation is determined at compile time based on the
+platform.
+
+There are three different implementations for spin locks in WiredTiger under the same interface:
+ - An implementation using GCC's builtin atomic operations.
+ - An implementation using a POSIX based mutex.
+ - An implementation for MSVC.
+Again specific implementation is determined at compile time based on the platform and compile-time
+configurations. WiredTiger has its own read-write lock implementation. It is based on atomic
+operations and condition variables.
+
+@subsection lock-usage Lock usage
+A correctly used lock takes the lock for each access to the shared memory it guards. WiredTiger
+sometimes deviates from this model and reads variables outside the context of the relevant lock.
+Such instances are only permitted if reading inconsistent data is sufficient for the use case.
+
+@section generations Generations
+
+Shared object, including the page index, WT_REFs, etc. have their lifecycles managed using
+the WiredTiger generations system. In general when a reader thread accesses an object, that could be
+concurrently freed, it enters a "generation". After entering that generation it accesses the object.
+A generation is a monotonically increasing integer which denotes times at which a given object must
+exist.
+
+Writer threads are then able to replace that object but cannot free it until the reader thread has
+left or released the generation. Instead the writer thread adds the object to a "stash" and it marks
+the stashed object with the generation indicating the point at which it was added to the stash. Once
+all reader in the system do not have a corresponding generation which is older than or equal to the
+stashed object's generation it can be freed.
+
+@section atomics Atomics
+
+As previously mentioned WiredTiger already expects loads and stores to specific sizes of points
+to be atomic and prevent tearing. WiredTiger also relies on atomic RMW operations, these include:
+- Compare and Swap
+- Fetch add / sub
+- Add / sub fetch
+
+Currently WiredTiger is undergoing a transition where all shared memory variable access will be
+wrapped by an atomic function, this will enable TSan to run on the code base. To facilitate this
+transition the atomic load and store macros have been added to WiredTiger, these macros placate
+TSan. To prevent them from introducing a performance regression they utilize relaxed consistency.
```

Comment:
```suggestion
transition atomic load and store macros have been added to WiredTiger. To prevent them from introducing a performance regression they utilize relaxed consistency.
```

Reply:
This was deleted.